internal/hammer/loadtest/workers.go

```diff
@@ -231,3 +273,178 @@ func (w *LogWriter) Kill() {
 		w.cancel()
 	}
 }
+
+// NewMMDVerifier creates a MMDVerifier.
+func NewMMDVerifier(tracker *client.LogStateTracker, errChan chan<- error, leafMMDChan chan LeafMMD, mmdDuration time.Duration) *MMDVerifier {
+	return &MMDVerifier{
+		tracker:     tracker,
+		errChan:     errChan,
+		leafMMDChan: leafMMDChan,
+		mmdDuration: mmdDuration,
+	}
+}
+
+// MMDVerifier verifies the signed timestamp against the MMD policy for newly
+// added entries by performing inclusion proof.
+type MMDVerifier struct {
+	tracker     *client.LogStateTracker
+	errChan     chan<- error
+	leafMMDChan chan LeafMMD
+	mmdDuration time.Duration
+	cancel      func()
+}
+
+// Run runs the MMD verifier. This should be called in a goroutine.
+func (v *MMDVerifier) Run(ctx context.Context) {
+	if v.cancel != nil {
+		panic("MMDVerifier was ran multiple times")
+	}
+	ctx, v.cancel = context.WithCancel(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		var leafMMD *LeafMMD
+		for {
+			if leafMMD == nil {
+				mmd, ok := <-v.leafMMDChan
+				if !ok {
+					break
+				}
+				leafMMD = &mmd
+			}
+
+			// Retry if the leaf is not yet integrated into the log.
+			if leafMMD.index >= v.tracker.LatestConsistent.Size {
+				// Verify MMD timestamp.
+				sec := int64(leafMMD.timestamp / 1000)
+				nsec := int64((leafMMD.timestamp % 1000) * 1000000)
+				if time.Unix(sec, nsec).Add(v.mmdDuration).Before(time.Now()) {
+					v.errChan <- fmt.Errorf("leaf index %d MMD violation at %d", leafMMD.index, leafMMD.timestamp)
+				}
+
+				break
+			}
+
+			// A new proof builder is required because v.tracker.ProofBuilder
+			// is not thread safe.
+			size := v.tracker.LatestConsistent.Size
+			hash := v.tracker.LatestConsistent.Hash
```

Comment:
I strongly encourage reading `v.tracker.LatestConsistent` to a variable early on, and then using it through a single run through the loops here. As it stands, the tracker could update between these two calls and you'd save the old size and the new hash, which would be totally broken.

Reply:
Good catch. Updated to use `v.tracker.LatestConsistent`.